File: tagbag/tag_analyzer.py
from .tag_base import TagBase
from .ollama_client import OllamaClient


from pathlib import Path
from typing import Iterable, List, Sequence


class TagAnalyzer(TagBase):
    """
    Analyze all tags across files using an LLM prompt.
    Prints the model output to stdout.
    """

    @staticmethod
    def execute(args: Sequence[str]) -> None:
        """
        args[0]: directory path
        args[1] optional: model name
        args[2] optional: prompt file name
        """
        dir_path = args[0]
        model = args[1] if len(args) >= 2 else ""
        prompt_file = args[2] if len(args) >= 3 else ""

        file_names = TagBase.get_tag_files(dir_path)

        content: List[str] = []
        TagAnalyzer._load_tags_from_files(file_names, content)

        print("Tags in files:")
        for c in content:
            print(c)

        summary = TagAnalyzer._analyze(content, model, prompt_file)
        for line in summary.splitlines():
            print(line)

    @staticmethod
    def _load_tags_from_files(file_names: Iterable[str], tokens: List[str]) -> None:
        for name in file_names:
            tokens.extend(Path(name).read_text(encoding="utf-8").splitlines())

    @staticmethod
    def _analyze(content: List[str], model: str, prompt_file_name: str) -> str:
        client = OllamaClient(model=model)
        try:
            lines = "\n".join(content)
            prompt_prefix = Path(prompt_file_name).read_text(encoding="utf-8") if prompt_file_name else ""
            prompt = f"{prompt_prefix}{lines}"
            return client.generate(prompt)
        except FileNotFoundError:
            return ""

File: tagbag/TagAnalyzer.cs

using Microsoft.Extensions.AI;
using Microsoft.VisualBasic;
using OllamaSharp;

namespace NhlStenden.TagBag;

/// <summary>
/// Represents a class that analyzes tags in files using AI models.
/// </summary>
public class TagAnalyzer : TagBase
{
    /// <summary>
    /// Executes the tag analyzing process on the specified files.
    /// </summary>
    /// <param name="args">A list of arguments where the first argument is the directory path.</param>
    public static void Execute(IList<string> args)
    {
        var fileNames = GetTagFiles(args[0]);
        var model = args.Count >= 2 ? args[1] : "";
        var promptFileName = args.Count >= 3 ? args[2] : "";

        var content = new List<string>();
        LoadTagsFromFiles(fileNames, content);

        Console.WriteLine("Tags in files:");
        foreach (var c in content)
            Console.WriteLine(c);

        var sum = Analyze(content, model, promptFileName);
        var lines = new List<string>(sum.Split(["\r\n", "\n", "\r"], StringSplitOptions.None));
        foreach (var line in lines)
            Console.WriteLine(line);
    }

    /// <summary>
    /// Reads all tags from multiple files into a list.
    /// </summary>
    /// <param name="fileNames">An array of file paths to process.</param>
    /// <param name="tokens">A list to collect the tags.</param>
    static void LoadTagsFromFiles(string[] fileNames, List<string> tokens)
    {
        foreach (var fileName in fileNames)
            tokens.AddRange(File.ReadAllLines(fileName));
    }

    /// <summary>
    /// Uses the specified AI model to analyze the given tags.
    /// </summary>
    /// <param name="content">The list of tags to process.</param>
    /// <param name="model">The name of the AI model to use for transformation.</param>
    /// <param name="promptFileName">The name of the prompt file to use for transformation.</param>
    /// <returns>The AI-generated analysis text, or an empty string if an error occurs.</returns>
    static string Analyze(List<string> content, string model, string promptFileName)
    {
        var client = new OllamaApiClient("http://localhost:11434", model);

        try
        {
            var lines = string.Join(Environment.NewLine, content);
            var prompt = string.Join(Environment.NewLine, File.ReadAllLines(promptFileName).ToList());
            prompt += lines;
            var r = Task.Run(() => client.GetResponseAsync(prompt)).GetAwaiter().GetResult();
            return r.Text;
        }
        catch (HttpRequestException)
        {
            return "";
        }
    }
}